File: Assets/Typescript~/src/Shared/UserInput/Drivers/KeyboardDriver.ts
import { Signal, SignalPriority } from "Shared/Util/Signal";
import { KeySignal } from "./Signals/KeySignal";
import { OnUpdate } from "Shared/Util/Timer";

export class KeyboardDriver {
	// public readonly KeyDown = new Signal<[key: KeySignal]>();
	// public readonly KeyUp = new Signal<[key: KeySignal]>();

	private readonly keyDownSignals = new Map<KeyCode, Signal<[key: KeySignal]>>();
	private readonly keyUpSignals = new Map<KeyCode, Signal<[key: KeySignal]>>();

	public readonly AnyKeyDownSignal = new Signal<[key: KeySignal]>();
	public readonly AnyKeyUpSignal = new Signal<[key: KeySignal]>();

	private static inst: KeyboardDriver;

	private constructor() {
		// UserInputService.InputProxy.OnKeyPressEvent((key, isDown) => {
		// 	if (isDown) {
		// 		this.KeyDown.Fire(new KeySignal(key));
		// 	} else {
		// 		this.KeyUp.Fire(new KeySignal(key));
		// 	}
		// });

		OnUpdate.Connect(() => {
			Profiler.BeginSample("PollForKeys");
			for (const [key, keyDownSignal] of this.keyDownSignals) {
				if (Input.GetKeyDown(key)) {
					const event = new KeySignal(key);
					this.AnyKeyDownSignal.Fire(event);
					if (event.IsCancelled()) continue;
					keyDownSignal.Fire(event);
				}
			}
			for (const [key, keyUpSignal] of this.keyUpSignals) {
				if (Input.GetKeyUp(key)) {
					const event = new KeySignal(key);
					this.AnyKeyUpSignal.Fire(event);
					if (event.IsCancelled()) continue;
					keyUpSignal.Fire(event);
				}
			}
			Profiler.EndSample();
		});
	}

	public OnKeyDown(key: KeyCode, callback: (event: KeySignal) => void, priority?: SignalPriority) {
		let signal = this.keyDownSignals.get(key);
		if (signal === undefined) {
			signal = new Signal();
			this.keyDownSignals.set(key, signal);
		}

		const disconnect = priority ? signal.ConnectWithPriority(priority, callback) : signal.Connect(callback);
		return () => {
			disconnect();

			// No more connections; remove the signal:
			if (!signal!.HasConnections()) {
				signal!.Destroy();
				this.keyDownSignals.delete(key);
			}
		};
	}

	public OnKeyUp(key: KeyCode, callback: (event: KeySignal) => void, priority?: SignalPriority) {
		let signal = this.keyUpSignals.get(key);
		if (signal === undefined) {
			signal = new Signal();
			this.keyUpSignals.set(key, signal);
		}

		const disconnect = priority ? signal.ConnectWithPriority(priority, callback) : signal.Connect(callback);
		return () => {
			disconnect();

			// No more connections; remove the signal:
			if (!signal!.HasConnections()) {
				signal!.Destroy();
				this.keyUpSignals.delete(key);
			}
		};
	}

	public IsKeyDown(key: KeyCode) {
		// return UserInputService.InputProxy.IsKeyDown(key);
		return Input.GetKey(key) || Input.GetKeyDown(key);
	}

	/** **NOTE:** Internal only. Use `Keyboard` class instead. */
	public static instance() {
		return (this.inst ??= new KeyboardDriver());
	}
}

File: Assets/Bundles/Shared/Resources/TS/UserInput/Drivers/KeyboardDriver.lua
-- Compiled with unity-ts v2.1.0-75
local Signal = require("Shared/TS/Util/Signal").Signal
local KeySignal = require("Shared/TS/UserInput/Drivers/Signals/KeySignal").KeySignal
local KeyboardDriver
do
	KeyboardDriver = setmetatable({}, {
		__tostring = function()
			return "KeyboardDriver"
		end,
	})
	KeyboardDriver.__index = KeyboardDriver
	function KeyboardDriver.new(...)
		local self = setmetatable({}, KeyboardDriver)
		return self:constructor(...) or self
	end
	function KeyboardDriver:constructor()
		self.KeyDown = Signal.new()
		self.KeyUp = Signal.new()
		UserInputService.InputProxy:OnKeyPressEvent(function(key, isDown)
			if isDown then
				self.KeyDown:Fire(KeySignal.new(key))
			else
				self.KeyUp:Fire(KeySignal.new(key))
			end
		end)
	end
	function KeyboardDriver:IsKeyDown(key)
		return UserInputService.InputProxy:IsKeyDown(key)
	end
	function KeyboardDriver:instance()
		if self.inst == nil then
			self.inst = KeyboardDriver.new()
		end
		return self.inst
	end
end
return {
	KeyboardDriver = KeyboardDriver,
}
-- ----------------------------------
-- ----------------------------------
